shiny/render/_dataframe.py

```diff
@@ -0,0 +1,320 @@
+    ):
+        """
+        Holds the data and options for a ``shiny.render.data_frame`` output, for a
+        spreadsheet-like view.
+
+        Parameters
+        ----------
+        data
+            A pandas `DataFrame` object, or any object that has a `.to_pandas()` method
+            (e.g., a Polars data frame or Arrow table).
+        width
+            A width for the data grid to occupy, in CSS units. The default is `fit-content`,
+            which sets the grid's width according to its contents. Set this to `100%` to use
+            the maximum available horizontal space.
+        height
+            A _maximum_ amount of vertical space for the data grid to occupy, in CSS units.
+            If there are more rows than can fit in this space, the grid will scroll.
+        summary
+            If `True` (the default), shows a message like "Viewing rows 1 through 10 of 20"
+            below the grid when not all of the rows are being shown. If `False`, the message
+            is not displayed. You can also specify a string template to customize the
+            message, for example: `"Viendo filas {start} a {end} de {total}"`.
+        row_selection_mode
+            Use `"none"` to disable row selection, `"single"` to allow a single row to be
+            selected at a time, and `"multi-toggle"` to allow multiple rows to be selected
+            by clicking on them individually.
+
+        Returns
+        -------
+        :
+            An object suitable for being returned from a `@render.data_frame`-decorated
+            output function.
+
+        See Also
+        --------
+        ~shiny.ui.output_data_frame
+        ~shiny.render.data_frame
+        """
+        import pandas as pd
+
+        self.data: pd.DataFrame = cast(
+            pd.DataFrame,
+            cast_to_pandas(
+                data,
+                "The DataGrid() constructor didn't expect a 'data' argument of type",
+            ),
+        )
+
+        self.width = width
+        self.height = height
+        self.summary = summary
+        self.row_selection_mode = row_selection_mode
+
+    def to_payload(self) -> object:
+        res: dict[str, Any] = json.loads(
+            # {index: [index], columns: [columns], data: [values]}
+            self.data.to_json(  # pyright: ignore[reportUnknownMemberType]
+                None, orient="split"
+            )
+        )
+        res["options"] = dict(
+            width=self.width,
+            height=self.height,
+            summary=self.summary,
+            row_selection_mode=self.row_selection_mode,
+            style="table",
+        )
+        return res
+
+
+# It would be nice to specify DataGridResult to be something like:
+#   Union[pandas.DataFrame, <protocol with .to_pandas()>]
+# However, if we did that, we'd have to import pandas at load time, which adds
+# a nontrivial amount of overhead. So for now, we're just using `object`.
+DataGridResult = Union[None, object, DataGrid]
```

Comment:
I believe you can make this work if you add

```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import pandas as pd
```

And then here:

```suggestion
DataGridResult = Union[None, "pd.DataFrame", DataGrid]
```

Note the quotes around `pd.DataFrame`. That's so that Python doesn't try to evaluate `pd.DataFrame` at load time. But the type checkers are smart enough to know that types in quotes are meant to be treated as types and not string literals.